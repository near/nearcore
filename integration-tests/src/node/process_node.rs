use std::ops::ControlFlow;
use std::path::PathBuf;
use std::process::{Child, Command};
use std::sync::Arc;
use std::time::Duration;
use std::{env, thread};

use rand::Rng;
use tracing::error;

use near_chain_configs::Genesis;
use near_crypto::{InMemorySigner, KeyType, Signer};
use near_primitives::types::AccountId;
use nearcore::config::NearConfig;

use crate::node::Node;
use crate::user::rpc_user::RpcUser;
use crate::user::User;
use actix::System;
use near_jsonrpc_client::new_client;
use near_network::test_utils::wait_or_timeout;

pub enum ProcessNodeState {
    Stopped,
    Running(Child),
}

pub struct ProcessNode {
    pub work_dir: PathBuf,
    pub config: NearConfig,
    pub state: ProcessNodeState,
    pub signer: Arc<InMemorySigner>,
}

impl Node for ProcessNode {
    fn genesis(&self) -> &Genesis {
        &self.config.genesis
    }

    fn account_id(&self) -> Option<AccountId> {
        self.config.validator_signer.as_ref().map(|vs| vs.validator_id().clone())
    }

    fn start(&mut self) {
        match self.state {
            ProcessNodeState::Stopped => {
                std::env::set_var("ADVERSARY_CONSENT", "1");
                let child =
                    self.get_start_node_command().spawn().expect("start node command failed");
                self.state = ProcessNodeState::Running(child);
                let client_addr = format!("http://{}", self.config.rpc_addr().unwrap());
                thread::sleep(Duration::from_secs(3));
                near_actix_test_utils::run_actix(async move {
                    wait_or_timeout(1000, 30000, || async {
                        let res = new_client(&client_addr).status().await;
                        if res.is_err() {
                            return ControlFlow::Continue(());
                        }
                        ControlFlow::Break(())
                    })
                    .await
                    .unwrap();
                    System::current().stop()
                });
            }
            ProcessNodeState::Running(_) => panic!("Node is already running"),
        }
    }

    fn kill(&mut self) {
        match self.state {
            ProcessNodeState::Running(ref mut child) => {
                child.kill().expect("kill failed");
                thread::sleep(Duration::from_secs(1));
                self.state = ProcessNodeState::Stopped;
            }
            ProcessNodeState::Stopped => panic!("Invalid state"),
        }
    }

    fn signer(&self) -> Arc<dyn Signer> {
        self.signer.clone()
    }

    fn is_running(&self) -> bool {
        match self.state {
            ProcessNodeState::Stopped => false,
            ProcessNodeState::Running(_) => true,
        }
    }

    fn user(&self) -> Box<dyn User> {
        let account_id = self.signer.account_id.clone();
        Box::new(RpcUser::new(&self.config.rpc_addr().unwrap(), account_id, self.signer.clone()))
    }

    fn as_process_ref(&self) -> &ProcessNode {
        self
    }

    fn as_process_mut(&mut self) -> &mut ProcessNode {
        self
    }
}

impl ProcessNode {
    /// Side effect: reset_storage
    pub fn new(config: NearConfig) -> ProcessNode {
        let mut rng = rand::thread_rng();
        let work_dir = env::temp_dir().join(format!("process_node_{}", rng.gen::<u64>()));
        let signer = Arc::new(InMemorySigner::from_seed(
            config.validator_signer.as_ref().unwrap().validator_id().clone(),
            KeyType::ED25519,
            config.validator_signer.as_ref().unwrap().validator_id().as_ref(),
        ));
        let result = ProcessNode { config, work_dir, state: ProcessNodeState::Stopped, signer };
        result.reset_storage();
        result
    }

    /// Clear storage directory and run keygen
    pub fn reset_storage(&self) {
        Command::new("rm").arg("-r").arg(&self.work_dir).spawn().unwrap().wait().unwrap();
        self.config.save_to_dir(&self.work_dir);
    }

    /// Side effect: writes chain spec file
    pub fn get_start_node_command(&self) -> Command {
        if std::env::var("NIGHTLY_RUNNER").is_err() {
            let mut command = Command::new("cargo");
            command.args(&["run", "-p", "neard"]);
            #[cfg(feature = "nightly_protocol")]
            command.args(&["--features", "nightly_protocol"]);
            #[cfg(feature = "nightly")]
            command.args(&["--features", "nightly"]);
            command.args(&["--bin", "neard", "--", "--home"]);
            command.arg(&self.work_dir);
            command.arg("run");
            command
        } else {
            let mut command = Command::new("target/debug/neard");
            command.arg("--home");
            command.arg(&self.work_dir);
            command.arg("run");
            command
        }
    }
}

impl Drop for ProcessNode {
    fn drop(&mut self) {
        match self.state {
            ProcessNodeState::Running(ref mut child) => {
                let _ = child.kill().map_err(|_| error!("child process died"));
                std::fs::remove_dir_all(&self.work_dir).unwrap();
            }
            ProcessNodeState::Stopped => {}
        }
    }
}
