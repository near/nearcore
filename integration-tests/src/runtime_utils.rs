use std::collections::HashSet;

use near_chain_configs::Genesis;
use near_primitives::shard_layout::ShardUId;
use near_primitives::state_record::{state_record_to_account_id, StateRecord};
use near_primitives::types::AccountId;
use near_primitives::types::StateRoot;
use near_store::genesis::GenesisStateApplier;
use near_store::test_utils::create_tries_complex;
use near_store::{ShardTries, TrieUpdate};
use nearcore::config::GenesisExt;
use node_runtime::config::RuntimeConfig;
use node_runtime::{state_viewer::TrieViewer, Runtime};
use testlib::runtime_utils::{add_test_contract, alice_account, bob_account};

pub const TEST_SHARD_UID: ShardUId = ShardUId { version: 1, shard_id: 0 };

pub fn get_runtime_and_trie() -> (Runtime, ShardTries, StateRoot) {
    let mut genesis = Genesis::test_sharded_new_version(
        vec![alice_account(), bob_account(), "carol.near".parse().unwrap()],
        3,
        vec![3],
    );
    add_test_contract(&mut genesis, &"test.contract".parse().unwrap());
    get_runtime_and_trie_from_genesis(&genesis)
}

pub fn get_test_trie_viewer() -> (TrieViewer, TrieUpdate) {
    let (_, tries, root) = get_runtime_and_trie();
    let trie_viewer = TrieViewer::default();
    let state_update = tries.new_trie_update(TEST_SHARD_UID, root);
    (trie_viewer, state_update)
}

pub fn get_runtime_and_trie_from_genesis(genesis: &Genesis) -> (Runtime, ShardTries, StateRoot) {
    let shard_layout = &genesis.config.shard_layout;
    let tries = create_tries_complex(shard_layout.version(), shard_layout.num_shards());
    let runtime = Runtime::new();
    let mut account_ids: HashSet<AccountId> = HashSet::new();
    genesis.for_each_record(|record: &StateRecord| {
        account_ids.insert(state_record_to_account_id(record).clone());
    });
    let writers = std::sync::atomic::AtomicUsize::new(0);
    let storage_usage_config = &RuntimeConfig::test().fees.storage_usage_config;
    let genesis_root = GenesisStateApplier::apply(
        &writers,
        tries.clone(),
        0,
        &genesis
            .config
            .validators
            .iter()
            .map(|account_info| {
                (
                    account_info.account_id.clone(),
                    account_info.public_key.clone(),
                    account_info.amount,
                )
            })
            .collect::<Vec<_>>(),
        storage_usage_config,
        genesis,
        account_ids,
    );
    (runtime, tries, genesis_root)
}
