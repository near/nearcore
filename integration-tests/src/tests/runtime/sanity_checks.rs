use crate::node::{Node, RuntimeNode};
use near_chain_configs::Genesis;
use near_primitives::config::ExtCosts;
use near_primitives::runtime::config::RuntimeConfig;
use near_primitives::runtime::config_store::RuntimeConfigStore;
use near_primitives::serialize::to_base64;
use near_primitives::types::AccountId;
use near_primitives::version::PROTOCOL_VERSION;
use near_primitives::views::{
    CostGasUsed, ExecutionOutcomeWithIdView, ExecutionStatusView, FinalExecutionStatus,
};
use nearcore::config::GenesisExt;
use std::collections::HashSet;
use std::mem::size_of;
use testlib::runtime_utils::{add_test_contract, alice_account, bob_account};

/// Initial balance used in tests.
const TESTING_INIT_BALANCE: u128 = 1_000_000_000 * NEAR_BASE;

/// One NEAR, divisible by 10^24.
const NEAR_BASE: u128 = 1_000_000_000_000_000_000_000_000;

/// Max prepaid amount of gas.
const MAX_GAS: u64 = 300_000_000_000_000;

/// Costs whose amount of `gas_used` may depend on environmental factors.
///
/// In particular, compiling our test contract with different versions of
/// compiler can lead to slightly different WASM output.
const NONDETERMINISTIC_COSTS: [&str; 2] = ["CONTRACT_LOADING_BYTES", "WASM_INSTRUCTION"];

fn is_nondeterministic_cost(cost: &str) -> bool {
    NONDETERMINISTIC_COSTS.iter().find(|&&ndt_cost| ndt_cost == cost).is_some()
}

fn test_contract_account() -> AccountId {
    format!("test-contract.{}", alice_account().as_str()).parse().unwrap()
}

fn setup_runtime_node_with_contract(wasm_binary: &[u8]) -> RuntimeNode {
    // Create a `RuntimeNode`. Load `RuntimeConfig` from `RuntimeConfigStore`
    // to ensure we are using the latest configuration.
    let mut genesis =
        Genesis::test(vec![alice_account(), bob_account(), "carol.near".parse().unwrap()], 3);
    add_test_contract(&mut genesis, &alice_account());
    add_test_contract(&mut genesis, &bob_account());
    let runtime_config_store = RuntimeConfigStore::new(None);
    let runtime_config = runtime_config_store.get_config(PROTOCOL_VERSION);
    let node = RuntimeNode::new_from_genesis_and_config(
        &alice_account(),
        genesis,
        RuntimeConfig::clone(runtime_config),
    );

    let account_id = node.account_id().unwrap();
    let node_user = node.user();
    let tx_result = node_user
        .create_account(
            account_id,
            test_contract_account(),
            node.signer().public_key(),
            TESTING_INIT_BALANCE / 2,
        )
        .unwrap();
    assert_eq!(tx_result.status, FinalExecutionStatus::SuccessValue(Vec::new()));
    assert_eq!(tx_result.receipts_outcome.len(), 2);

    let tx_result =
        node_user.deploy_contract(test_contract_account(), wasm_binary.to_vec()).unwrap();
    assert_eq!(tx_result.status, FinalExecutionStatus::SuccessValue(Vec::new()));
    assert_eq!(tx_result.receipts_outcome.len(), 1);

    node
}

fn get_receipts_status_with_clear_hash(
    outcomes: &[ExecutionOutcomeWithIdView],
) -> Vec<ExecutionStatusView> {
    outcomes
        .iter()
        .map(|outcome| {
            match outcome.outcome.status {
                ExecutionStatusView::SuccessReceiptId(_) => {
                    // We donâ€™t control the hash of the receipt so clear it.
                    ExecutionStatusView::SuccessReceiptId(Default::default())
                }
                ref status => status.clone(),
            }
        })
        .collect::<Vec<_>>()
}

/// Calls method `sanity_check` on `test-contract-rs` and verifies that the
/// resulting gas profile matches expectations.
///
/// This test intends to catch accidental configuration changes, see #4961.
#[test]
fn test_cost_sanity() {
    let test_contract = if cfg!(feature = "nightly") {
        near_test_contracts::nightly_rs_contract()
    } else {
        near_test_contracts::rs_contract()
    };
    let node = setup_runtime_node_with_contract(test_contract);

    let args = format!(
        r#"{{
            "contract_code": {:?},
            "method_name": "main",
            "method_args": "",
            "validator_id": {:?}
        }}"#,
        to_base64(near_test_contracts::trivial_contract()),
        bob_account().as_str()
    );
    eprintln!("{args}");

    let res = node
        .user()
        .function_call(
            alice_account(),
            test_contract_account(),
            "sanity_check",
            args.into_bytes(),
            MAX_GAS,
            0,
        )
        .unwrap();
    assert_eq!(res.status, FinalExecutionStatus::SuccessValue(Vec::new()));
    assert_eq!(res.transaction_outcome.outcome.metadata.gas_profile, None);

    let receipts_status = get_receipts_status_with_clear_hash(&res.receipts_outcome);
    insta::assert_yaml_snapshot!("receipts_status", receipts_status);

    let receipts_gas_profile = res
        .receipts_outcome
        .iter()
        .map(|outcome| outcome.outcome.metadata.gas_profile.as_ref().unwrap())
        .map(|gas_profile| {
            gas_profile
                .iter()
                .cloned()
                .map(|cost| {
                    if is_nondeterministic_cost(&cost.cost) {
                        // Ignore `gas_used` of nondeterministic costs.
                        CostGasUsed { gas_used: 0, ..cost }
                    } else {
                        cost
                    }
                })
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();

    insta::assert_debug_snapshot!(
        if cfg!(feature = "nightly") {
            "receipts_gas_profile_nightly"
        } else {
            "receipts_gas_profile"
        },
        receipts_gas_profile
    );
}

/// Verifies the sanity of nondeterministic costs using a trivial contract.
///
/// Some costs are nondeterministic since they depend on environmental factors,
/// see [`NONDETERMINISTIC_COSTS`]. For a trivial contract, however, there are
/// no such differences expected.
#[test]
fn test_cost_sanity_nondeterministic() {
    let contract = near_test_contracts::wat_contract(
        r#"(module (func (export "main") (i32.const 92) (drop)))"#,
    );
    let node = setup_runtime_node_with_contract(&contract);
    let res = node
        .user()
        .function_call(alice_account(), test_contract_account(), "main", vec![], MAX_GAS, 0)
        .unwrap();
    assert_eq!(res.status, FinalExecutionStatus::SuccessValue(Vec::new()));
    assert_eq!(res.transaction_outcome.outcome.metadata.gas_profile, None);

    let receipts_status = get_receipts_status_with_clear_hash(&res.receipts_outcome);
    insta::assert_yaml_snapshot!("receipts_status_nondeterministic", receipts_status);

    let receipts_gas_profile = res
        .receipts_outcome
        .iter()
        .map(|outcome| outcome.outcome.metadata.gas_profile.as_ref().unwrap())
        .collect::<Vec<_>>();

    insta::assert_debug_snapshot!(
        if cfg!(feature = "nightly") {
            "receipts_gas_profile_nondeterministic_nightly"
        } else {
            "receipts_gas_profile_nondeterministic"
        },
        receipts_gas_profile
    );

    // Verify that all nondeterministic costs are covered.
    let all_costs = {
        let all_costs = receipts_gas_profile
            .iter()
            .flat_map(|gas_profile| gas_profile.iter().map(|cost| cost.cost.as_str()));
        HashSet::from_iter(all_costs)
    };
    let ndt_costs = HashSet::from(NONDETERMINISTIC_COSTS);
    let missing_costs = ndt_costs.difference(&all_costs).collect::<Vec<_>>();
    assert!(
        missing_costs.is_empty(),
        "some nondeterministic costs are not covered: {:?}",
        missing_costs,
    );
}

/// Verifies the operation of host function `used_gas` according to
/// [gas instrumentation].
///
/// [gas instrumentation]: https://nomicon.io/RuntimeSpec/Preparation#gas-instrumentation
#[test]
fn test_sanity_used_gas() {
    let node = setup_runtime_node_with_contract(&contract_sanity_check_used_gas());
    let res = node
        .user()
        .function_call(alice_account(), test_contract_account(), "main", vec![], MAX_GAS, 0)
        .unwrap();

    let num_return_values = 4;
    let returned_bytes = match res.status {
        FinalExecutionStatus::SuccessValue(v) => v,
        _ => panic!("Unexpected status: {:?}", res.status),
    };
    assert_eq!(returned_bytes.len(), num_return_values * size_of::<u64>());

    let used_gas = stdx::as_chunks_exact::<{ size_of::<u64>() }, _>(&returned_bytes)
        .unwrap()
        .iter()
        .map(|bytes| u64::from_le_bytes(*bytes))
        .collect::<Vec<_>>();

    let runtime_config = node.client.read().unwrap().runtime_config.clone();
    let base_cost = runtime_config.wasm_config.ext_costs.gas_cost(ExtCosts::base);
    let op_cost = match runtime_config.wasm_config.limit_config.contract_prepare_version {
        // In old implementations of preparation, all of the contained instructions are paid
        // for upfront when entering a new metered block,
        //
        // It fails to be precise in that it does not consider calls to `used_gas` as
        // breaking up a metered block and so none of the gas cost to execute instructions between
        // calls to this function will be observable from within wasm code.
        //
        // In this test we account for this by setting `op_cost` to zero, but if future tests
        // change test WASM in significant ways, this approach may become incorrect.
        near_primitives::config::ContractPrepareVersion::V0
        | near_primitives::config::ContractPrepareVersion::V1 => 0,
        // Gas accounting is precise and instructions executed between calls to the side-effectful
        // `used_gas` host function calls will be observbable.
        near_primitives::config::ContractPrepareVersion::V2 => {
            u64::from(runtime_config.wasm_config.regular_op_cost)
        }
    };

    // Executing `used_gas` costs `base_cost` plus an instruction to execute the `call` itself.
    // When executing `used_gas` twice within a metered block, the returned values should differ by
    // that amount.
    assert_eq!(used_gas[1] - used_gas[0], base_cost + op_cost);
    // Between these two observations additional arithmetics have been executed.
    assert_eq!(used_gas[2] - used_gas[1], base_cost + 8 * op_cost);
    assert!(used_gas[3] - used_gas[2] > base_cost);
}

/// Returns a contract which calls host function `used_gas` multiple times, both
/// within the same [metered block] and in different metered blocks. The value
/// returned by the contract is an array of values returned by calls of
/// `used_gas`.
///
/// This contract is written in `wat` to avoid depending on the output generated
/// by a compiler (e.g. `rustc`).
///
/// [metered block]: https://nomicon.io/RuntimeSpec/Preparation#gas-instrumentation
fn contract_sanity_check_used_gas() -> Vec<u8> {
    wat::parse_str(
        r#"
          (module
            (type $t0 (func (result i64)))
            (type $t1 (func (param i64 i64)))
            (type $t2 (func))

            (import "env" "used_gas" (func $env.used_gas (type $t0)))
            (import "env" "value_return" (func $env.value_return (type $t1)))

            (memory 1)

            (func $main (export "main") (type $t2)
              (local $used_0 i64) (local $used_1 i64) (local $used_2 i64) (local $used_3 i64)

              ;; Call used_gas twice in metered block, without instructions in between.
              (call $env.used_gas)
              (call $env.used_gas)
              (local.set $used_1)
              (local.set $used_0)

              ;; In the same metered block, call used_gas again after executing other
              ;; instructions.
              (i64.add
                (local.get $used_0)
                (i64.const 1))
              drop
              nop

              (call $env.used_gas)
              (local.set $used_2)

              ;; Push a new metered block on the stack via br_if. The condition is false,
              ;; so we will _not_ branch out of the block.
              (block $b0
                (br_if $b0
                  (i64.eq
                    (local.get $used_0)
                    (i64.const 0)))
                  (local.set $used_3
                    call $env.used_gas)
                  nop ;; ensure there is more than used_gas to pay for in this block
              )

              ;; Prepare bytes passed to value_return.
              (i64.store
                (i32.const 0)
                (local.get $used_0))
              (i64.store
                (i32.const 8)
                (local.get $used_1))
              (i64.store
                (i32.const 16)
                (local.get $used_2))
              (i64.store
                (i32.const 24)
                (local.get $used_3))

              (call $env.value_return
                (i64.const 32)
                (i64.extend_i32_u
                  (i32.const 0)))
            )
          )
 "#,
    )
    .unwrap()
}
