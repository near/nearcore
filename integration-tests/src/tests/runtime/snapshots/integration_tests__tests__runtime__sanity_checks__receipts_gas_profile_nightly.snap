---
source: integration-tests/src/tests/runtime/sanity_checks.rs
expression: receipts_gas_profile
---
[
    [
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "ADD_FULL_ACCESS_KEY",
            gas_used: 101765125000,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "ADD_FUNCTION_CALL_KEY_BASE",
            gas_used: 102217625000,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "ADD_FUNCTION_CALL_KEY_BYTE",
            gas_used: 9626655,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "CREATE_ACCOUNT",
            gas_used: 7700000000000,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "DELETE_ACCOUNT",
            gas_used: 147489000000,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "DELETE_KEY",
            gas_used: 94946625000,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "DEPLOY_CONTRACT_BASE",
            gas_used: 184765750000,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "DEPLOY_CONTRACT_BYTE",
            gas_used: 231641966,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "FUNCTION_CALL_BASE",
            gas_used: 20878753500000,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "FUNCTION_CALL_BYTE",
            gas_used: 207941862,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "NEW_ACTION_RECEIPT",
            gas_used: 1480548358496,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "STAKE",
            gas_used: 141715687500,
        },
        CostGasUsed {
            cost_category: "ACTION_COST",
            cost: "TRANSFER",
            gas_used: 230246125000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "ALT_BN128_G1_MULTIEXP_BASE",
            gas_used: 713000000000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "ALT_BN128_G1_MULTIEXP_ELEMENT",
            gas_used: 320000000000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "ALT_BN128_G1_SUM_BASE",
            gas_used: 3000000000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "ALT_BN128_G1_SUM_ELEMENT",
            gas_used: 5000000000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "ALT_BN128_PAIRING_CHECK_BASE",
            gas_used: 9686000000000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "ALT_BN128_PAIRING_CHECK_ELEMENT",
            gas_used: 5102000000000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "BASE",
            gas_used: 17209927215,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BASE",
            gas_used: 35445963,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BYTES",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "LOG_BASE",
            gas_used: 7086626100,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "LOG_BYTE",
            gas_used: 131987910,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "PROMISE_AND_BASE",
            gas_used: 1465013400,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "PROMISE_AND_PER_PROMISE",
            gas_used: 87234816,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "PROMISE_RETURN",
            gas_used: 560152386,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "READ_CACHED_TRIE_NODE",
            gas_used: 4560000000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "READ_MEMORY_BASE",
            gas_used: 182690424000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "READ_MEMORY_BYTE",
            gas_used: 4744063584,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "READ_REGISTER_BASE",
            gas_used: 7551495558,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "READ_REGISTER_BYTE",
            gas_used: 25034748,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "RIPEMD160_BASE",
            gas_used: 853675086,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "RIPEMD160_BLOCK",
            gas_used: 680107584,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "SHA256_BASE",
            gas_used: 4540970250,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "SHA256_BYTE",
            gas_used: 120586755,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_HAS_KEY_BASE",
            gas_used: 108079793250,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_HAS_KEY_BYTE",
            gas_used: 277117605,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_READ_BASE",
            gas_used: 112713691500,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_READ_KEY_BYTE",
            gas_used: 278572797,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_READ_VALUE_BYTE",
            gas_used: 28055025,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_REMOVE_BASE",
            gas_used: 106946061000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_REMOVE_KEY_BYTE",
            gas_used: 343983456,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_REMOVE_RET_VALUE_BYTE",
            gas_used: 57657780,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_WRITE_BASE",
            gas_used: 128393472000,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_WRITE_EVICTED_BYTE",
            gas_used: 160586535,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_WRITE_KEY_BYTE",
            gas_used: 281931468,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "STORAGE_WRITE_VALUE_BYTE",
            gas_used: 310185390,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "TOUCHING_TRIE_NODE",
            gas_used: 32203911852,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "UTF16_DECODING_BASE",
            gas_used: 3543313050,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "UTF16_DECODING_BYTE",
            gas_used: 1635774930,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "UTF8_DECODING_BASE",
            gas_used: 46676685915,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "UTF8_DECODING_BYTE",
            gas_used: 98262621423,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "VALIDATOR_STAKE_BASE",
            gas_used: 911834726400,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "VALIDATOR_TOTAL_STAKE_BASE",
            gas_used: 911834726400,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WASM_INSTRUCTION",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WRITE_MEMORY_BASE",
            gas_used: 19626564027,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WRITE_MEMORY_BYTE",
            gas_used: 866159496,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WRITE_REGISTER_BASE",
            gas_used: 37251792318,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WRITE_REGISTER_BYTE",
            gas_used: 1904583564,
        },
    ],
    [
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "BASE",
            gas_used: 264768111,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BASE",
            gas_used: 35445963,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BYTES",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WASM_INSTRUCTION",
            gas_used: 0,
        },
    ],
    [],
    [
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "BASE",
            gas_used: 264768111,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BASE",
            gas_used: 35445963,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BYTES",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "READ_MEMORY_BASE",
            gas_used: 2609863200,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "READ_MEMORY_BYTE",
            gas_used: 11403999,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "UTF8_DECODING_BASE",
            gas_used: 3111779061,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "UTF8_DECODING_BYTE",
            gas_used: 874741437,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WASM_INSTRUCTION",
            gas_used: 0,
        },
    ],
    [],
    [
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BASE",
            gas_used: 35445963,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BYTES",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WASM_INSTRUCTION",
            gas_used: 0,
        },
    ],
    [],
    [
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BASE",
            gas_used: 35445963,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BYTES",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WASM_INSTRUCTION",
            gas_used: 0,
        },
    ],
    [],
    [
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BASE",
            gas_used: 35445963,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BYTES",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WASM_INSTRUCTION",
            gas_used: 0,
        },
    ],
    [],
    [],
    [],
    [],
    [],
    [
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BASE",
            gas_used: 70891926,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BYTES",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WASM_INSTRUCTION",
            gas_used: 0,
        },
    ],
    [],
    [],
    [],
    [],
    [
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BASE",
            gas_used: 35445963,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BYTES",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WASM_INSTRUCTION",
            gas_used: 0,
        },
    ],
    [],
    [
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "BASE",
            gas_used: 529536222,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BASE",
            gas_used: 35445963,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "CONTRACT_LOADING_BYTES",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WASM_INSTRUCTION",
            gas_used: 0,
        },
        CostGasUsed {
            cost_category: "WASM_HOST_COST",
            cost: "WRITE_REGISTER_BASE",
            gas_used: 2865522486,
        },
    ],
    [],
    [],
]
